### ISSUE TITLE
[Question] or perhaps [Bug] pulling does not update chached value

### ISSUE THREAD

**Issue by gprsnlty** [ztalbot2000/homebridge-cmd4#86]
I my config I set `"fetch": Cached,` and set `"polling": [{"characteristic": "on", "interval": 40, "timeout": 8000}]` but it seams like polling does not have any effect.

I Expected polling to update the Cached On value every interval (here 40sec) but it seems like the cached value does not get updated.

Do I miss understand something or is this unexpected behaviour?

Below I try to explain my understanding of the Readme. But if you don't want to read all this in short: It would be great if you could help me and explain the three options to me.
I really like your plugin and if this is a stupid question I hope I don't annoy you too much. 

**tl;dr:**

_{ "fetch": "Always" } - As before, Always fetch characteristic value. ( Default )_

This polls the characteristic On value whenever there is a request e.g. I open the Home app? And If `"polling":[{"char... "}]` is set as well it does not change anything? 

_{ "fetch": "Cached" } - Never fetch characteristic value. Use cached value. The cached value would have to be updated through polling._

This never polls the On value but uses the Cached On value when there is a request. And `"polling":[{"char... "}]` updates the Cached On value every interval (here 40sec)? But if `"polling":[{"char... "}]` is not set the cached value uses Get/Set cached (according to config.min.json line 24). what is the Get/Set cached do I need to handle those in my script like a Parameter of Get/Set sate ('SetCached' 'name' 'value' / 'GetCached' 'name' or sth like this).
Also I do not understand  (config.min.json line 52) _// * Developrs who use their own scripts must understand this directive. As Cached only Polled characteristics states are retrieved immediately through the state_cmd._

_{ "fetch": "Polled" } - Polled characteristics act like before, "Always". Non polled characteristic values are fetched from cache._

This polls chracteristics wich are inside a "polling":[{"char... "}] every time there is a request (like it was set to Always). But characteristics wich are not in a `"polling":[{"char... "}]` are retrived from cache? And there fore never get updated since they are not polled at all.

**Comment by ztalbot2000** [comment]
Hi,

  Yeah, I thought about this last night that the cached value does not get
updated. Theoretically it should not matter, at least so I thought of at
the time.  I'm working on something at the moment.

I'll fix it within the day.

TTFN,
John

**Comment by ztalbot2000** [comment]
Hi,

  Correction,  it gets updated on the next get, not the set. But if the get
is a cached get then it would be the old value.  Weirdly, wrapping your
head around how this could happen is confusing; but I am working on it.

**Comment by gprsnlty** [comment]
Cool! Thank you very much that you care about such little issues!

**Comment by ztalbot2000** [comment]
I was working on it as you noticed.  I just published Cmd4 v3.0.15 that addresses this.
For cached characteristics that are not polled, but you update them from HomeKit, they now get saved for the next cached get.  
Someho I think this may not be good enough in that like fetch, their should be three options. Call the directive stache.  It would be stache Always, Polled and Cached.  What do you think?

I'll close this for now as I think this is what you need.

**Comment by gprsnlty** [comment]
I just did a quick test so I might be overseeing something.
But when I now push the button in HomeKit the homebridge output is
`[Cmd4] setCachedvalue accTypeEnumIndex:( 105 )-"On" function for: android value: false`
and the button changes its state.
But cmd4 does not send a Setvalue 'Set' 'name' 'value' and does not run the script.

**Comment by ztalbot2000** [comment]
This is where things are interesting, because if you want it sent, then
fetch should be "Polled" or "Always".  Otherwise why call it cached?

TTFN,
John

**Comment by gprsnlty** [comment]
ah ok I feel stupid Im sorry I did not realize that that is what cache is used for.
I apologize for wasting your time here thank you for your help!

**Comment by ztalbot2000** [comment]
Don't feel stupid, because in ways I still think you have a point.  There
is just so many things to think of.  Another update may come tomorrow.   I
just need a clear head.



TTFN,
John

**Comment by ztalbot2000** [comment]
Hi,

   I still have this in the back of my mind.  I added to the Cmd4 Portal an
image of the Fetch Flow Diagram
<https://ztalbot2000.github.io/homebridge-cmd4/images/fetchFlowDiagram.png> for
Cmd4. I did this to get it straight in my own mind. It is easy to
understand that after the "Set" for "Cached" or "Always" what happens. When
fetch is set to "Polled" though. A set that talks to the device, if it was
successful I think maybe it should set the internal value, which it
currently does not.
Consider that the "Set" in this case happens successfully, but the next get
does not happen for a long time. The internal state of the characteristic
is incorrect, but does it matter since the "Get" would have to be for the
cached value? Is there any case where the "Get" is a cached get and not an
uncached "Get". This is the question I have to ask myself, because if the
"Set" was to the device, then so should be the "Get".
I think it is best to store the successful set of the cached value. At
least this is what I'm thinking at the moment.
If you are wondering why I am asking you, sometimes just writing out the
problem gives me insight.  If you would like to comment, you are welcome
to. This may be more for myself.

ttfn,
John Talbot

**Comment by ztalbot2000** [comment]
Just a note on this one.  I shall be releasing 3.3 soon.  Polling calls updateValue as it is supposed to for background changes and while notifications still work, it is possibly more reliably so as this is what is recommended by Homebridge.
